core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,332 @@
+use super::{hasher, AdviceSetError, Felt, FieldElement, Word};
+// use math::fields::f64::BaseElement;
+use std::collections::BTreeMap;
+
+// MERKLE PATH SET
+// ================================================================================================
+
+/// A set of Merkle paths.
+///
+/// This struct is intended to be used as one of the variants of the MerkleSet enum.
+#[derive(Clone, Debug)]
+pub struct MerklePathSet {
+    root: Word,
+    total_depth: u32,
+    paths: BTreeMap<u64, Vec<Word>>,
+}
+
+impl MerklePathSet {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an empty MerklePathSet.
+    pub fn new(depth: u32) -> Result<Self, AdviceSetError> {
+        let root = [Felt::ZERO; 4];
+        let paths = BTreeMap::<u64, Vec<Word>>::new();
+
+        Ok(Self {
+            root,
+            total_depth: depth,
+            paths,
+        })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds the Merkle Path to the paths map.
+    /// index -- index of the node, which path we want to add
+    /// hash -- hash of the node itself
+    /// path -- Merkle Path of the node
+    pub fn add_path(
+        &mut self,
+        index: u64,
+        value: Word,
+        path: Vec<Word>,
+    ) -> Result<(), AdviceSetError> {
+        let depth = (path.len() + 1) as u32;
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidPathLength(self.total_depth, depth));
+        }
+
+        // actual number of node in tree
+        let pos = 2u64.pow(self.total_depth) + index;
+
+        // index of the leaf path in map. We store paths of neighboring leaves in one key-value pair
+        let half_pos = (pos / 2) as u64;
+
+        // Now we ignoring adding path to the node with existing path
+        if self.paths.get(&half_pos) == None {
```

Comment:
For completeness, if there is a path with the same index already in the map, it might be good to check that the new path is the same as the one in the map, and if not, return an error.